File: MinimumSegmentTree.java
class MinimumSegmentTree{
    private int lengthOfArray;
    private long[] arr;
    private long[] segmentArray;
    private int heightOfSegmentTree;

    public MinimumSegmentTree(long[] a){
        this.lengthOfArray = a.length;
        this.arr = new long[lengthOfArray];
        for(int i=0;i<a.length;i++) arr[i] = a[i];
        this.heightOfSegmentTree = (int)Math.ceil(Math.log(lengthOfArray)/Math.log(2));
        int lengthOfSegmentArray = 2*(1<<heightOfSegmentTree)-1;
        this.segmentArray = new long[lengthOfSegmentArray];
        minimumSegmentTreeConstructorUtil(0,lengthOfArray-1,0);
    }

    private long minimumSegmentTreeConstructorUtil(int segmentStart, int segmentEnd, int segmentIndex){
        if(segmentEnd == segmentStart) return segmentArray[segmentIndex] = arr[segmentStart];
        int mid = getMid(segmentStart,segmentEnd);
        return segmentArray[segmentIndex] = Math.min(minimumSegmentTreeConstructorUtil(segmentStart,mid,segmentIndex*2+1),
                minimumSegmentTreeConstructorUtil(mid+1,segmentEnd,segmentIndex*2+2));
    }

    private int getMid(int segmentStart, int segmentEnd){
        return (segmentStart+segmentEnd)>>1;
    }

    public long getMinimumInclusiveOfRange(int left, int right){
        return getMinimumUtil(0,lengthOfArray-1,left,right,0);
    }
    private long getMinimumUtil(int segmentStart, int segmentEnd, int queryStart, int queryEnd, int segmentIndex){
        if(queryStart<=segmentStart && segmentEnd<=queryEnd) return segmentArray[segmentIndex];
        if(queryEnd<segmentStart || segmentEnd<queryStart) return Long.MAX_VALUE;
        int mid = getMid(segmentStart,segmentEnd);
        return Math.min(getMinimumUtil(segmentStart,mid,queryStart,queryEnd,segmentIndex*2+1),
                getMinimumUtil(mid+1,segmentEnd,queryStart,queryEnd,segmentIndex*2+2));
    }
    public void update(int index, long value){
        updateConstructorUtil(0,lengthOfArray-1,0,index,value);
    }
    private long updateConstructorUtil(int segmentStart, int segmentEnd, int segmentIndex, int index, long newValue){
        if(index>segmentEnd || index<segmentStart) return segmentArray[segmentIndex];
        if(segmentStart==index && segmentEnd==index) return segmentArray[segmentIndex]=newValue;
        int mid = getMid(segmentStart,segmentEnd);
        return segmentArray[segmentIndex]=Math.min(updateConstructorUtil(segmentStart,mid,2*segmentIndex+1,index,newValue),
                updateConstructorUtil(mid+1,segmentEnd,2*segmentIndex+2,index,newValue));
    }
}

File: Segment Trees/MinimumSegmentTree/MinimumSegmentTree.cpp
class MinimumSegmentTree{
    private:
    int n,*sa,h,st_len;
    int constructor(int ss, int se, int si, int a[]){
        if(ss == se) return sa[si] = a[ss];
        int mid = (ss+se)>>1;
        return sa[si] = min(constructor(ss,mid,2*si+1,a),constructor(mid+1,se,2*si+2,a));
    }
    int minimum(int ss, int se, int si, int qs, int qe){
        if(qe<ss || se<qs) return LONG_LONG_MAX;
        if(qs<=ss && qe>=se) return sa[si];
        int mid = (ss+se)>>1;
        return min(minimum(ss,mid,2*si+1,qs,qe),minimum(mid+1,se,2*si+2,qs,qe));
    }
    int update(int ss, int se, int si, int index, int value){
        if(index<ss || index>se) return sa[si];
        if(index == ss && index == se) return sa[si] = value;
        int mid = (ss+se) >> 1;
        return sa[si] = min(update(ss,mid,2*si+1,index,value),update(mid+1,se,2*si+2,index,value));
    }
    public:
    MinimumSegmentTree(int a[], int a_len){
        n = a_len;
        h = ceil(log(n)/log(2));
        st_len = (1<<(h+1))-1;
        sa = (int*)malloc(st_len*sizeof(int));
        constructor(0,n-1,0,a);
    }
    int getMinimum(int left, int right){
        if(left > right) return 0;
        return minimum(0,n-1,0,left,right);
    }
    void updateIndex(int index, int value){
        update(0,n-1,0,index,value);
    }
};
